src/librustc_incremental/persist/load.rs

```diff
@@ -168,35 +170,40 @@ pub fn load_dep_graph(sess: &Session) ->
-        time_ext(time_passes, None, "background load prev dep-graph", move || {
-            match load_data(report_incremental_info, &path) {
-                LoadResult::DataOutOfDate => LoadResult::DataOutOfDate,
-                LoadResult::Error { message } => LoadResult::Error { message },
-                LoadResult::Ok { data: (bytes, start_pos) } => {
-
-                    let mut decoder = Decoder::new(&bytes, start_pos);
-                    let prev_commandline_args_hash = u64::decode(&mut decoder)
-                        .expect("Error reading commandline arg hash from cached dep-graph");
-
-                    if prev_commandline_args_hash != expected_hash {
-                        if report_incremental_info {
-                            println!("[incremental] completely ignoring cache because of \
-                                    differing commandline arguments");
-                        }
-                        // We can't reuse the cache, purge it.
-                        debug!("load_dep_graph_new: differing commandline arg hashes");
-
-                        // No need to do any further work
-                        return LoadResult::DataOutOfDate;
-                    }
+    let load = move || {
+        match load_data(report_incremental_info, &path) {
+            LoadResult::DataOutOfDate => LoadResult::DataOutOfDate,
+            LoadResult::Error { message } => LoadResult::Error { message },
+            LoadResult::Ok { data: (bytes, start_pos) } => {
+
+                let mut decoder = Decoder::new(&bytes, start_pos);
+                let prev_commandline_args_hash = u64::decode(&mut decoder)
+                    .expect("Error reading commandline arg hash from cached dep-graph");
 
-                    let dep_graph = SerializedDepGraph::decode(&mut decoder)
-                        .expect("Error reading cached dep-graph");
+                if prev_commandline_args_hash != expected_hash {
+                    if report_incremental_info {
+                        println!("[incremental] completely ignoring cache because of \
+                                differing commandline arguments");
+                    }
+                    // We can't reuse the cache, purge it.
+                    debug!("load_dep_graph_new: differing commandline arg hashes");
 
-                    LoadResult::Ok { data: PreviousDepGraph::new(dep_graph) }
+                    // No need to do any further work
+                    return LoadResult::DataOutOfDate;
                 }
+
+                let dep_graph = SerializedDepGraph::decode(&mut decoder)
+                    .expect("Error reading cached dep-graph");
+
+                LoadResult::Ok { data: PreviousDepGraph::new(dep_graph) }
             }
+        }
+    };
+
+    MaybeAsync::Async(std::thread::spawn(move || {
+        time_ext(time_passes, None, "background load prev dep-graph", move || {
+            panic::catch_unwind(load).unwrap_or_else(
```

Comment:
Do we need to catch the unwind? Can't we just handle this on thread.join() since we're paying that cost anyway?

Reply:
That's a good idea! I will do some testing and see if that works.

Reply:
Yes, that's a good suggestion.